TACOptimizations/ControlFlowGraph.cs

```diff
@@ -78,13 +92,44 @@ public ControlFlowGraph(List<BasicBlock> basicBlocks)
                         break;
                 }
             }
+
+            DFS();
+        }
+
+        public void DFS()
+        {
+            _dfst = new List<(int, int)>();
+
+            var c = _basicBlocks.Count;
+            _lrn = new List<int>(c);
+            _nlr = new List<int>(c);
+            _dfn = new int[c];
+
+            var used = new bool[c];
+
+            Action<int> dfs = null;
+            dfs = (int vertex) =>
+            {
+                used[vertex] = true;
+                _nlr.Add(vertex);
+                foreach ((var v, _) in _children[vertex])
+                    if (!used[v])
+                    {
+                        _dfst.Add((vertex, v));
+                        dfs(v);
+                    }
+                _lrn.Add(vertex);
+                _dfn[vertex] = --c;
+            };
+
+            dfs(0);
         }
 
         public int VertexOf(BasicBlock block) => _blockToVertex[block];
         public List<BasicBlock> GetCurrentBasicBlocks() => _basicBlocks.ToList();
 
-        public List<(int, BasicBlock)> GetChildrenBasicBlocks(int vertex) => _children[vertex];
+        public List<(int, BasicBlock)> GetChildrenBasicBlocks(int vertex) => _children[vertex].ToList();
```

Comment:
Зачем `ToList()`? Это ведь и так список.
```suggestion
        public List<(int, BasicBlock)> GetChildrenBasicBlocks(int vertex) => _children[vertex];
```

Reply:
Копия
Чтобы кто-нибудь наверняка не сломал граф своими изменениями

Reply:
А почему копия? Если хочется запретить изменения графа со стороны, то это нужно отражать в интерфейсе
```suggestion
        public IReadOnlyList<(int, BasicBlock)> GetChildrenBasicBlocks(int vertex) => _children[vertex].AsReadOnly();
```
аналогично для `GetParentsBasicBlocks` и `GetCurrentBasicBlocks`.

Reply:
Я как-то использовал эти readonly
они не полностью покрывают list
ну и сейчас менять тип такое себе

Reply:
Что значит не полностью?
хотелось бы услышать что-то большее, чем "такое себе", хоть какие-то аргументы против. Если банально лень, то с этим я и сам справлюсь.
Вместо подразумевания можно как раз заюзать другой тип, чтобы никто не думал про изменение, а просто так отдавать копию нет смысла.
Уж если начал фиксить "уязвимости", то сделай это нормально.

Reply:
то и значит, не все методы доступны, вероятно нужно подключать еще че то
чтобы никто не думал - как раз и нефиг типы менять
копия как раз самое адекватное решение
еще и вероятно куда эффективнее чем readonly обертка
чет хз про что спор

Reply:
Всё верно, в IReadOnlyList внезапно не будут доступны методы, которые модифицируют список 🤣
Вообще-то наоборот, создание копии это линейная операция, а возврат `AsReadOnly()` -- константная. Ну и совсем не ясно, зачем на каждый запрос создавать копию...
Ещё раз, спор о том, что если хочется запретить модификацию драгоценных списков, то нужно просто возвращать коллекции, которые не получится поменять.

Reply:
Про модифицирующие я ничего не говорил
документацию почитал смотрю
тогда посмотри и сравни методы
https://docs.microsoft.com/ru-ru/dotnet/api/system.collections.generic.ilist-1
https://docs.microsoft.com/ru-ru/dotnet/api/system.collections.generic.ireadonlylist-1

придется кастить обратно в List, нафига это нужно
еще и менять написанное

"линейная операция" - там больше двух элементов не может быть

Reply:
А тебе все эти отсутствующие методы нужны или это так, лишь бы были?
Если очень хочется, то `IndexOf` и `Contains` это обычный цикл по коллекции, ума много не надо, чтобы это написать.

Что ты кастить собрался? Ничего кастить не нужно.

Менять написанное придётся ровно в 4 (четырёх) местах, заменить одно слово на другое. Extensions можно не трогать, они уже не нужны и мы это сейчас удалим.

Ты первый начал про эффективность говорить, не я. Больше двух детей не может быть, родителей может быть сколько угодно, базовых блоков тоже может быть сколько угодно.

Reply:
я не понимаю зачем усложнять
разве что, со мной так интересно беседовать

Reply:
Если для тебя замена `List` на `IReadOnlyList` это такое огромное осложнение, то очень печально, ведь это лучше отражает неизменяемость графа, которую ты хотел сохранить.

Постарался объяснить, как мог, чем это лучше выкидывания копии, от тебя же я ничего не услышал, кроме нежелания поменять 4 слова в других файлах, и какого-то желания сохранить методы, которые нигде не используются и не предполагаются к использованию (будет надо, тогда и добавим).

Reply:
Это мажорное изменение
Ты понимаешь, что сейчас уже пишутся задачи на этом графе
Зачем лишний раз потом править что-то

Reply:
Мне никаких проблем
Ну единственное я этим с пятницы смогу заняться лишь

Reply:
> Ты понимаешь, что сейчас уже пишутся задачи на этом графе

Первое: не вижу ни одной команды, которая бы писала уже какие-то новые задачи (посмотрел в Network Graph репозитория).
Второе: изменить одно слово они смогут, не волнуйся. Если кто-то вдруг почему-то менял или хочет менять эти списки, то пусть копию создаст и играется с ними сколько угодно.